File: leetcode/tree/binary_tree_maximum_path.py
from common import TreeNode

class Solution():
    def __init__(self):
        self.maxs = -9999999999
    
    def maxPathSum(self, root: TreeNode) -> int:
        def helper(root: TreeNode) -> int:
            ls = 0
            rs = 0
            if not root.left and not root.right:
                self.maxs = max(self.maxs, root.val)
                return root.val
            if root.left:
                ls = helper(root.left)
            if root.right:
                rs = helper(root.right)
            
            if ls >= 0 and rs >= 0:
                self.maxs = max(self.maxs, root.val + ls + rs)
                return root.val + max(ls, rs)
            elif ls < 0 and rs >= 0:
                self.maxs = max(self.maxs, root.val + rs)
                return root.val + rs
            elif ls >= 0 and rs < 0:
                self.maxs = max(self.maxs, root.val + ls)
                return root.val + ls
            else:
                self.maxs = max(self.maxs, root.val)
                return root.val
        helper(root)
        return self.maxs

if __name__ == "__main__":
    root = TreeNode(-10)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)
    print(Solution().maxPathSum(root))

File: leetcode/tree/binary_tree_maximum_path.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxPathSum(root *TreeNode) int {
	var maxs = -99999999
	helper(root, &maxs)
	return maxs
}

func helper(root *TreeNode, maxs *int) int {
	ls := 0
	rs := 0
	if root.Left == nil && root.Right == nil {
		*maxs = max(*maxs, root.Val)
		return root.Val
	}
	if root.Left != nil {
		ls = helper(root.Left, maxs)
	}
	if root.Right != nil {
		rs = helper(root.Right, maxs)
	}
	if ls >= 0 && rs >= 0 {
		*maxs = max(*maxs, root.Val+ls+rs)
		return root.Val + max(ls, rs)
	} else if ls >= 0 && rs < 0 {
		*maxs = max(*maxs, root.Val+ls)
		return root.Val + ls
	} else if ls < 0 && rs >= 0 {
		*maxs = max(*maxs, root.Val+rs)
		return root.Val + rs
	} else {
		*maxs = max(*maxs, root.Val)
		return root.Val
	}
}

func max(t1, t2 int) int {
	if t1 < t2 {
		return t2
	}
	return t1
}

func main() {
	root := &TreeNode{Val: -10, Left: nil, Right: nil}
	root.Left = &TreeNode{Val: 9, Left: nil, Right: nil}
	root.Right = &TreeNode{Val: 20, Left: nil, Right: nil}
	root.Right.Left = &TreeNode{Val: 15, Left: nil, Right: nil}
	root.Right.Right = &TreeNode{Val: 7, Left: nil, Right: nil}
	fmt.Println(maxPathSum(root))
}
